app/src/main/java/org/oppia/app/utility/Logger.kt

```diff
@@ -0,0 +1,122 @@
+  private val GLOBAL_LOG_LEVEL = LogLevel.VERBOSE
+  private val LOG_DIRECTORY =File(context.filesDir, "oppia_app.log")
+
+  private enum class LogLevel private constructor(val logLevel: Int) {
+    VERBOSE(Log.VERBOSE),
+    DEBUG(Log.DEBUG),
+    INFO(Log.INFO),
+    WARNING(Log.WARN),
+    ERROR(Log.ERROR),
+    ASSERT(Log.ASSERT)
+  }
+
+  /** Logs a verbose message with the specified tag.*/
+  fun v(tag: String, msg: String) {
+    write(LogLevel.VERBOSE, tag, msg)
+  }
+
+  /** Logs a verbose message with the specified tag, message and exception.*/
+  fun v(tag: String, msg: String, tr: Throwable) {
+    write(LogLevel.VERBOSE, tag, msg, tr)
+  }
+
+  /** Logs a debug message with the specified tag*/
+  fun d(tag: String, msg: String) {
+    write(LogLevel.DEBUG, tag, msg)
+  }
+
+  /** Logs a debug message with the specified tag, message and exception.*/
+  fun d(tag: String, msg: String, tr: Throwable) {
+    write(LogLevel.DEBUG, tag, msg, tr)
+  }
+
+  /** Logs a info message with the specified tag.*/
+  fun i(tag: String, msg: String) {
+    write(LogLevel.INFO, tag, msg)
+  }
+
+  /** Logs a info message with the specified tag, message and exception.*/
+  fun i(tag: String, msg: String, tr: Throwable) {
+    write(LogLevel.INFO, tag, msg, tr)
+  }
+
+  /** Logs a warn message with the specified tag.*/
+  fun w(tag: String, msg: String) {
+    write(LogLevel.WARNING, tag, msg)
+  }
+
+  /** Logs a warn message with the specified tag, message and exception.*/
+  fun w(tag: String, msg: String, tr: Throwable) {
+    write(LogLevel.WARNING, tag, msg, tr)
+  }
+
+  /** Logs a error message with the specified tag.*/
+  fun e(tag: String, msg: String) {
+    write(LogLevel.ERROR, tag, msg)
+  }
+
+  /** Logs a error message with the specified tag, message and exception.*/
+  fun e(tag: String, msg: String, tr: Throwable) {
+    write(LogLevel.ERROR, tag, msg, tr)
+  }
+
+  private fun isLogEnable(logLevel: LogLevel): Boolean {
+    return GLOBAL_LOG_LEVEL.logLevel < logLevel.logLevel
+  }
+
+  private fun write(logLevel: LogLevel, tag: String, log: String) {
+    if (isLogEnable(logLevel) && ENABLE_CONSOLE_LOG) {
+      Log.println(logLevel.logLevel, tag, log)
+    }
+    if (isLogEnable(logLevel) && ENABLE_FILE_LOG) {
+      val msg = Calendar.getInstance().time.toString() + "\t" + logLevel.name + "/" + tag + ": " + log
+      write(msg)
+    }
+  }
+
+  private fun write(logLevel: LogLevel, tag: String, log: String, tr: Throwable) {
+    if (isLogEnable(logLevel) && ENABLE_CONSOLE_LOG) {
+      Log.println(logLevel.logLevel, tag, log + "\n" + Log.getStackTraceString(tr))
+    }
+    if (isLogEnable(logLevel) && ENABLE_FILE_LOG) {
+      val msg =
+        Calendar.getInstance().time.toString() + "\t" + logLevel.name + "/" + tag + ": " + log + "\n" + Log.getStackTraceString(
+          tr
+        )
+      write(msg)
+    }
+  }
+
+  private fun write(text: String) {
+    val ps = PrintStream(LOG_DIRECTORY)
+    System.setOut(ps)
```

Comment:
Ah, this doesn't seem like a good practice since it's replacing default system behavior. We really ought to write to our print stream directly, and save it as a long-lived field.

However, this is still performing I/O on the main thread, and we shouldn't do that. We should use a Kotlin coroutine with the blocking dispatcher to perform that operation, instead.

Reply:
I have changed the way of writing to file. Also Kotlin coroutines is been introduced to handle IO operations